### PR TITLE
PRC feedback - CURIE-like prefixes

### PR DESCRIPTION
**Pull request by denis-yuen**
There is a section in the PRC document that suggests the use of CURIE-like IDs. 
Our current schema says little about the format for IDs, simply that they should be unique per implementation of a schema. (i.e. IDs uniquely identify tools in a particular registry implementation). 

The PRC feedback is the following:
```
“Id”: value described as “scoped identifier” => but what about requiring CURIEs / properly versioned ids?
Would add value to the service
Core property for all the services for GA4GH
Prefix should be clarified,

“Id” and “name” => possible clashes; so for “labeled ids“ following the { "id": "CURIE", "label": "some_string" } format?

Should tool id (https://github.com/ga4gh/tool-registry-service-schemas/blob/0d36f7199c1c32446127f242406c0214f19c597c/openapi/openapi.yaml#L621-L624) only allow [A-Za-z0-9.-_~] similar to DRS IDs (https://github.com/ga4gh/data-repository-service-schemas/blob/develop/docs/asciidoc/front_matter.adoc#drs-ids)
```

I think this is saying that CURIE like IDs should be mandated for all implementations of TRS resulting in IDs that look like `dockstore:12345` as opposed to the current example of `123456`  https://github.com/ga4gh/tool-registry-service-schemas/blob/2.0.0-beta.3/openapi/ga4gh-tool-discovery.yaml#L498 ) The last note seems to imply that if we go down this route, `:` should be a reserved character to distinguish between the prefix and the actual ID. 

An explanation about the second concern (possible classes for id and name) would be helpful. 
Thanks!

### CODE REVIEW

File: openapi/ga4gh-tool-discovery.yaml [hunk 1]
```diff
@@ -38,8 +38,8 @@ paths:
           required: true
           type: string
           description: >-
-            A unique identifier of the tool, scoped to this registry, for
-            example `123456`.
+            A unique identifier of the tool, scoped to this registry, with a CURIE-like prefix for
```
Comment (coverbeck) [inline openapi/ga4gh-tool-discovery.yaml]: I think the convention with `for example` is to have commas before and after. But you'll end up with a slightly unwieldy sentence.

File: openapi/openapi.yaml [hunk 2]
```diff
@@ -620,8 +620,16 @@ components:
...
           type: string
-          example: 123456
-          description: A unique identifier of the tool, scoped to this registry.
+          example: dockstore:123456
+          description: A unique identifier of the tool, scoped to this registry with a
+            CURIE-like prefix. See
+            https://github.com/ga4gh/tool-registry-service-schemas/blob/develop/registry.json
+            for the current list of prefixes and the registry that they
+            correspond to. TRS IDs are strings made up of uppercase and
+            lowercase letters, decimal digits, hypen, period, percent,
```
Comment (coverbeck) [inline openapi/openapi.yaml]: Typo in `hyphen`.

File: openapi/openapi.yaml [hunk 3]
```diff
@@ -620,8 +620,16 @@ components:
...
+            underscore and tilde [A-Za-z0-9.-%_~] to simplify parsing CURIE-like
+            prefixes. Special characters like slashes can be encoded to work
+            with these restrictions.
```
Comment (coverbeck) [inline openapi/openapi.yaml]: The more I think about it, I'm not sure, but I would lean towards `within` instead of `with`.

File: openapi/ga4gh-tool-discovery.yaml (hunk 1 above)
Comment (garyluu) [reply openapi/ga4gh-tool-discovery.yaml]: `x-example` property also works.  It should get automatically converted to `example` in openapi

**Comment by denis-yuen** [comment]
```
junjun_zhang  [35 minutes ago]
I think there is a reason for DRS to pick up `[A-Za-z0-9.-_~]`
junjun_zhang  [35 minutes ago]
and all other char should be encoded
denis.yuen  [35 minutes ago]
I was wondering about that, because I thought they would need `%` in order to encode stuff (edited)
junjun_zhang  [25 minutes ago]
https://github.com/ga4gh/data-repository-service-schemas/blob/release/drs-1.0.0/docs/asciidoc/front_matter.adoc#drs-ids
junjun_zhang  [23 minutes ago]
they referred to this: https://tools.ietf.org/html/rfc3986#section-2.3
denis.yuen  [22 minutes ago]
ok, I guess that means the colon must be encoded. Rats
denis.yuen  [19 minutes ago]
On the other hand https://tools.ietf.org/html/rfc3986#section-3.3 seems to say that a colon is fine as long as its not in the first path segment
junjun_zhang  [12 minutes ago]
yeah, I don't quite understand it, but I think DRS tries to follow this standard for it's ID.
junjun_zhang  [11 minutes ago]
I think if possible, all GA4GH standards should follow the same ID standard whatever it might be
junjun_zhang  [10 minutes ago]
similarly for `name` field although it may have different standard compared to ID field
denis.yuen  [10 minutes ago]
I'm not sure, it would be kind of gross to enforce CURIE prefixes but then say the colon always has to be URL encoded
junjun_zhang  [6 minutes ago]

  An example of CURIE syntax: [isbn:0393315703]

  The square brackets may be used to prevent ambiguities between CURIEs and regular URIs, yielding so-called safe CURIEs.

junjun_zhang  [6 minutes ago]
in the above example, I think the actual ID (uri) is still `0393315703` (edited)
junjun_zhang  [5 minutes ago]
`[isbn:0393315703]` is CURIEs way of that ID
junjun_zhang  [4 minutes ago]
basically, internally, there are two different fields, say: `id` and `namespace`
junjun_zhang  [3 minutes ago]
then put them together like: `[namespace:id]` to for CURIEs ID
junjun_zhang  [3 minutes ago]
that's my understanding
denis.yuen  [3 minutes ago]
What does that look like in terms of endpoints?
Is it `/api/ga4gh/v2/tools/isbn:0393315703/versions/1` or `/api/ga4gh/v2/tools/isbn%3A0393315703/versions/1` 
```

**Comment by denis-yuen** [comment]
Linking a bunch of tickets for further reading 
https://issues.jboss.org/browse/RESTEASY-1117 
https://github.com/swagger-api/swagger-js/pull/977 
https://bugs.launchpad.net/ubuntu/+source/gnome-terminal/+bug/377367

**Comment by denis-yuen** [comment]
Question from cloud workstream: this enforced a id format, and where would the global registry for id prefixes live (assumed in the registry.json in this repo)

David G: what happens with a private TRS, want to hide public knowledge of a prefix. 
Further discussion as to the wisdom of a URL-like CURIE format

**Comment by delagoya** [comment]
@susheel @ypriverol We would need support from Biocontainers and/or Elixir for finalizing any changes to spec.

**Comment by hmenager** [comment]
discussion on this point with @denis-yuen , my 2 cents here (@denis-yuen doesn't have strong feelings about this) :

- global identifiers are a good thing for reproducibility/FAIRness/etc
- issue with private TRS servers: it would be wise to require people to set up their own prefix.
- full URL vs ID:
 e.g.: imagine `https://dockstore.org/api/ga4gh/v2/tools/dockstore:12345` vs `dockstore:12345`.
rather should `dockstore:12345` resolve to `https://dockstore.org/api/ga4gh/v2/tools/12345` using a resolver service?
have there been similar discussions for other GA4GH e.g. DRS. @denis-yuen thinks there were some, but doesn't remember the outcome.

**Comment by rishidev** [comment]
PRC comments
- standard should recommend prefixes for private repos (optional), but mandatory for public. Allows for a resolver in the Networks setup. Having prefixes makes it easier to use for newcomers.
- ID is preferred method over the URL. More stable over time.

**Comment by mbaudis** [comment]
Regarding resolvers for CURIEs: We've run in some issues recently where even for "easy" ids there were incompatibilities (e.g. PubMed CURIEs are frequently `PMID:123456`, but identifiers.org only resolves them in the `pubmed:123456` format, currently).

This lead to the wish for having a tested/appropriate resolver solution documented in the spec., even if not explicitly requesting it. Kind of a middle ground between URL and CURIE, in that the more stable, compact, readable CURIE is used as the standard format, while documenting its use for accessing the endpoint.

**Comment by tetron** [comment]
In the RDF world, CURIEs just denote namespaces, they yield global identifiers by simple string substitution without requiring a resolver.  This requires knowing which namespaces are active for every document or context, but avoids the kind of problems @mbaudis mentions because (presumably) `PMID:` and `pubmed:` would just be aliases for the same global URI.

**Comment by mbaudis** [comment]
@tetron Sure. Therefore also the "provide a working example in the comments", instead of "coding in a resolver (list)". See recent discussion [here](https://github.com/phenopackets/phenopacket-schema/issues/177#issuecomment-514828134).

**Comment by denis-yuen** [comment]
For discussion, an alternative based on discussion with @hmenager 
https://github.com/ga4gh/tool-registry-service-schemas/pull/96

Open to ideas from @tetron and @mbaudis but I'm out of my element with RDF and CURIEs

**Comment by denis-yuen** [comment]
Agreed, closing
